controllers/btpoperator_controller.go

```diff
@@ -576,14 +631,64 @@ func (r *BtpOperatorReconciler) deleteCreationTimestamp(us ...*unstructured.Unst
 	}
 }
 
-func (r *BtpOperatorReconciler) setConfigMapValues(secret *corev1.Secret, u *unstructured.Unstructured) error {
-	return unstructured.SetNestedField(u.Object, string(secret.Data["cluster_id"]), "data", "CLUSTER_ID")
+func (r *BtpOperatorReconciler) setConfigMapValues(baseSecret *corev1.Secret, configMap *unstructured.Unstructured, resolvedNamespace string) error {
+	err := unstructured.SetNestedField(configMap.Object, resolvedNamespace, "data", "MANAGEMENT_NAMESPACE")
+	err = unstructured.SetNestedField(configMap.Object, resolvedNamespace, "data", "RELEASE_NAMESPACE")
+
+	err = unstructured.SetNestedField(configMap.Object, string(baseSecret.Data["cluster_id"]), "data", "CLUSTER_ID")
+
+	return err
+}
+
+func (r *BtpOperatorReconciler) checkIfOperatorNeedRestart(ctx context.Context, resolvedNamespace string, baseSecret *corev1.Secret, logger *logr.Logger) bool {
+	detectConfigMapChanges := func() bool {
+		currentConfigMap := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      ConfigName,
+				Namespace: ChartNamespace,
+			},
+		}
+		err := r.Get(context.TODO(), client.ObjectKeyFromObject(currentConfigMap), currentConfigMap)
+		if k8serrors.IsNotFound(err) {
+			logger.Info(fmt.Sprintf("ConfigMap %s in %s not found, so restart is needed", ConfigName, ChartNamespace))
+			return true
+		}
+
+		return !reflect.DeepEqual(baseSecret.Data["management_namespace"], []byte(currentConfigMap.Data["RELEASE_NAMESPACE"])) ||
+			!reflect.DeepEqual(baseSecret.Data["management_namespace"], []byte(currentConfigMap.Data["MANAGEMENT_NAMESPACE"]))
+	}
+
+	detectClusterIdChange := func() bool {
+		clusterIdSecret := corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      clusterIdSecretName,
+				Namespace: resolvedNamespace,
+			},
+		}
+		err := r.Client.Get(ctx, client.ObjectKeyFromObject(&clusterIdSecret), &clusterIdSecret)
+		if k8serrors.IsNotFound(err) {
+			logger.Info(fmt.Sprintf("secret %s in %s not found, so restart is needed", clusterIdSecretName, resolvedNamespace))
+			return true
+		}
+		return !reflect.DeepEqual(baseSecret.Data["cluster_id"], clusterIdSecret.Data["INITIAL_CLUSTER_ID"])
+	}
+
+	configMapChanged := detectConfigMapChanges()
+	logger.Info(fmt.Sprintf("is %s in %s changed? %t", ConfigName, ChartNamespace, configMapChanged))
+	clusterIdSecretChanged := detectClusterIdChange()
+	logger.Info(fmt.Sprintf("is %s %s changed? %t", clusterIdSecretName, resolvedNamespace, clusterIdSecretChanged))
+
+	return configMapChanged || clusterIdSecretChanged
 }
 
 func (r *BtpOperatorReconciler) setSecretValues(secret *corev1.Secret, u *unstructured.Unstructured) error {
 	for k := range secret.Data {
-		if err := unstructured.SetNestedField(u.Object, base64.StdEncoding.EncodeToString(secret.Data[k]), "data", k); err != nil {
-			return err
+		for _, r := range requiredSecretKeys {
```

Comment:
This is over-engineered.
We just need to change one value of the map.
I think no loop is needed, at least not two nested loops.

Reply:
We will discuss it later on.